### PR TITLE
Allow parsing the value of java.version sysprop

### LINKED ISSUE #43141: [TEST] SslMultiPortTests failing on 7.x
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+matrix-java-periodic/ES_BUILD_JAVA=openjdk12,ES_RUNTIME_JAVA=openjdk13,nodes=immutable&&linux&&docker/16/consoleText

https://gradle.com/s/zudrwwx3tyrgi

Relevant exception:

```
 java.lang.IllegalArgumentException: value
        at __randomizedtesting.SeedInfo.seed([E2E5B6E4A3D8F189:5242019AEA205556]:0)
        at org.elasticsearch.bootstrap.JavaVersion.parse(JavaVersion.java:48)
        at org.elasticsearch.xpack.security.transport.ssl.SslMultiPortTests.lambda$getProtocols$1(SslMultiPortTests.java:439)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:312)
        at org.elasticsearch.xpack.security.transport.ssl.SslMultiPortTests.getProtocols(SslMultiPortTests.java:439)
        at org.elasticsearch.xpack.security.transport.ssl.SslMultiPortTests.testThatTransportClientWithOnlyTruststoreCannotConnectToClientProfile(SslMultiPortTests.java:312)
```

The fix might be to use `java.specification.version` instead of `java.version` for parsing, or directly use `JavaVersion.current()`

### PR DESCRIPTION
**Pull request by jkakavas**
We often start testing with early access versions of new Java
versions and this have caused minor issues in our tests
(i.e. #43141) because the version string that the JVM reports
cannot be parsed as it ends with the string `-ea`.

This commit changes how we parse and compare Java versions to
allow correct parsing and comparison of the output of `java.version` 
system property that might include an additional alphanumeric
part after the version numbers (see [JEP 223[(https://openjdk.java.net/jeps/223)). 

It also changes a number of tests that would attempt to parse
`java.specification.version` in order to get the full version
of Java. `java.specification.version` only contains the major
version and is thus inappropriate when trying to compare against
a version that might contain a minor, patch or an early access
part. We know parse `java.version` that can be consistently
parsed.

Resolves #43141

### CODE REVIEW

**Comment by jkakavas** [comment]
@elasticmachine update branch

**Comment by davidkyle** [comment]
@jkakavas if this fixes #43141 please backport all the way to 7.3 as `SslMultiPortTests` are also failing there

**Comment by jkakavas** [comment]
> @jkakavas if this fixes #43141 please backport all the way to 7.3 as SslMultiPortTests are also failing there

It does, if es-core-infra likes this approach, I'll backport this to 7.3 too , otherwise I'll come up with a solution to handle the tests only

**Comment by rjernst** [comment]
Thanks for looking into this @jkakavas!

While I understand the need to handle `-ea` in parsing, I don't think we should treat EA version as special, in that we should not have the `isEarlyAccess()` flag. Long term, we should be using Runtime.Version, which is provided as part of the JDK. The portion of the version specification that allows `-ea` can have other things, it is the "$PRE" portion of `$VNUM(-$PRE)?\+$BUILD(-$OPT)?` in [JEP 322](https://openjdk.java.net/jeps/322).

The difference here with what you have implemented is we shouldn't try to discern any meaning out of `$PRE`. The specification allows it to be any alphanum string. For comparison logic, the javadocs state:

> A version with a pre-release identifier is always considered to be less than a version without one.  Pre-release identifiers are compared numerically when they consist only of digits, and lexicographically otherwise

So, I think we can get the same behavior you are after, but without having any "-ea" special string or comparison.

**Comment by jkakavas** [comment]
Got it , makes sense @rjernst. I didn't meant to attach any special meaning to the early access flag, `isEarlyAccess()` was just a utility method for the comparison part. I'll address this to make the comparison generic as dictated in JEP-322  and keep an eye for this to move to Runtime.Version when java 8 won't be supported.

**Comment by jkakavas** [comment]
Changed this to be able to handle the value of `java.version` (`$VNUM(\-$PRE)?`) and `java.specification.version` (`$VNUM`) as these are the use cases we currently have. In order to support fully parsing java.runtime.version , we'd have to replicate much of `Runtime.Version` logic and there currently doesn't seem to be a reason to do so.

**Comment by jkakavas** [comment]
No rush, just pinging for visibility @rjernst.

File: libs/core/src/main/java/org/elasticsearch/bootstrap/JavaVersion.java [hunk 1]
```diff
@@ -42,23 +45,38 @@ private JavaVersion(List<Integer> version) {
...
-            version.add(Integer.valueOf(component));
+        String[] parts = value.split("-");
+        String[] numericComponents;
+        if (parts.length == 1) {
+            numericComponents = value.split("\\.");
+        } else if (parts.length == 2) {
+            numericComponents = parts[0].split("\\.");
+            prePart = parts[1];
+        } else {
+            throw new IllegalArgumentException("value");
```
Comment (rjernst) [inline libs/core/src/main/java/org/elasticsearch/bootstrap/JavaVersion.java]: We should have a more descriptive failure message, which includes the actual value, not the literal string `value`. ;)

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by jkakavas** [comment]
@rjernst This wouldn't normally go into a patch release (7.3.1) but it does resolve the CI issues we see in 7.3 (https://github.com/elastic/elasticsearch/issues/43141) so I'm thinking I'd backport this to 7.3 too. Objections ?

**Comment by jkakavas** [comment]
Both ci/1 and ci/2 have completed successfully but the notification failed to reach :octocat: because of the service hiccups. 
```
18:53:20 Could not update commit status of the Pull Request on GitHub.
18:53:20 org.kohsuke.github.HttpException: Server returned HTTP response code: 500, message: 'Internal Server Error' for URL: https://api.github.com/repos/elastic/elasticsearch/statuses/...
```
ci/1: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/3323/console
ci/2 : https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-2/3302/console

**Pull request merged**